runtime/doc/diagnostic.txt

```diff
@@ -0,0 +1,616 @@
+
+                Parameters: ~
+                    {namespace}  number|nil The namespace to get diagnostics
+                                 from. If omitted, get diagnostics from all
+                                 namespaces. be `nil` only when {bufnr} is not
+                                 `nil` .
+                    {bufnr}      number|nil The buffer number to get
+                                 diagnistics from. If omitted, get diagnostics
+                                 from all buffers.
+
+                Return: ~
+                    table A list of diagnostic items |diagnostic-structure|.
+
+                                                  *vim.diagnostic.get_count()*
+get_count({namespace}, {bufnr}, {severity})
+                Get the number of diagnostics with the given severity.
+
+                Useful for showing diagnostic counts in statusline. eg: >
+
+                 function! DiagnosticStatus() abort
+                   let sl = ''
+                   let sl .= 'E: %{luaeval("vim.diagnostic.get_count(nil, 0, vim.diagnostic.severity.ERROR)")} '
+                   let sl .= 'W: %{luaeval("vim.diagnostic.get_count(nil, 0, vim.diagnostic.severity.WARN)")}'
+                   return sl
+                 endfunction
+<
+
+                Then simply add `%{DiagnosticStatus()}` to your statusline.
+
+                Parameters: ~
+                    {namespace}     number|nil Limit diagnostics to the given
+                                    namespace. If `nil` , retrieve diagnostics
+                                    from all namespaces.
+                    {bufnr}         number|nil The buffer number. Defaults to
+                                    the current buffer.
+                    {severity|nil}  Return diagnostics for the given severity
+                                    |vim.diagnostic.severity|. If omitted, all
+                                    diagnositcs are returned.
+
+                                       *vim.diagnostic.get_line_diagnostics()*
+get_line_diagnostics({bufnr}, {lnum}, {opts})
+                Get diagnostics by line.
+
+                Parameters: ~
+                    {namespace}  number|nil Limit diagnostics to the given
+                                 namespace. If `nil` , retrieve diagnostics
+                                 from all namespaces.
+                    {bufnr}      number|nil The buffer number. Defaults to the
+                                 current buffer.
+                    {lnum}       number|nil The line number. Defaults to the
+                                 current line.
+                    {opts}       table|nil A table with the following keys:
+                                 • namespace: (number) Limit diagnostics to
+                                   the given namespace.
+                                 • severity: (|vim.diagnostic.severity|) Only
+                                   return diagnostics with this severity.
```

Comment:
Given that it said "Overrides severity_limit." I would have expected it to behave like a limit, also given this is the behavior I'd expect as a _user_.   The expectation assumed here is that "exact match" is the common case and limit is some kind of special case, and I think it would be less surprising the other way around (alternatively `severity_limit` and `severity_only` and an error for a conflict to be neutral and explicit about behavior).

Reply:
I think the behavior (and the documentation broadly) is good; the documentation also matches the (simple) implementation where `severity` is checked first.

Two suggestions for improvement:
* switch the order, so that `limit` comes first; that makes the contrast clearer and gives context for the "overrides".
* if `severity` should be renamed, I'd suggest `severity_level` -- and use that consistently, also in the code (e.g., `filter_by_severity_level` rather than `filter_at_severity_limit` -- which I had to read three times before noticing the difference to `filter_by_severity_limit`...)

Reply:
And while we're at naming, I can't help but add my *ceterum censeo* that the inconsistent use of `diagnostic` and `diagnostics` is a mild but constant irritation...

If we do make a breaking change to the API (even with a compatibility layer to make it transparent), I'd love to take the opportunity to standardize on `vim.diagnostic**s**` everywhere!

Reply:
As promoting this to _top level_ core API will substantially increase the direct usage of this API among plugins and users (as opposed to indirectly by activating a LSP config), I do think it is worthwhile to do some polish at the API at the same time, even if this implies a breaking change for some existing usages (no opinion either way on a compat layer)

Reply:
Is `min_severity` clearer than `severity_limit`? I will also make `severity` and `min_severity` mutually exclusive; providing both is an error.

```
- severity: (|vim.diagnostic.severity|) Only return diagnostics with this severity. Conflicts with {min_severity}.
- min_severity: (|vim.diagnostic.severity|) Return diagnostics with this severity or worse. Conflicts with {severity}.
```

My only hangup is that `min` implies ordered numbering. In my mind "lower" = "better" so that e.g. "Warning" is "lower" than "Error". So `min_severity = WARN` would imply both warnings and errors.